community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/verification/DuplicateCheckStrategy.java

```diff
@@ -0,0 +1,274 @@
+
+            // We either have to find the first conflicting entry set element,
+            // or append one for the property we just fetched:
+            scan:
+            do
+            {
+                for ( int i = 0; i < entrySetSize; i++ )
+                {
+                    Object[] currentValues = (Object[])current.value[i];
+
+                    if ( current.nodeId[i] == StatementConstants.NO_SUCH_NODE )
+                    {
+                        current.value[i] = values;
+                        current.nodeId[i] = nodeId;
+                        if ( i == entrySetSize - 1 )
+                        {
+                            current.next = new EntrySet( entrySetSize );
+                        }
+                        break scan;
+                    }
+                    else if ( propertyValuesEqual( properties, currentValues ) )
+                    {
+                        throw new IndexEntryConflictException( current.nodeId[i], nodeId, currentValues );
+                    }
+                }
+                current = current.next;
+            }
+            while ( current != null );
+        }
+
+        @Override
+        void checkForDuplicate( Property property, Object propertyValue, long nodeId ) throws IndexEntryConflictException
+        {
+            EntrySet current = bucketEntrySet( propertyValue.hashCode(), entrySetSize );
+
+            // We either have to find the first conflicting entry set element,
+            // or append one for the property we just fetched:
+            scan:
+            do
+            {
+                for ( int i = 0; i < entrySetSize; i++ )
+                {
+                    Object value = current.value[i];
+
+                    if ( current.nodeId[i] == StatementConstants.NO_SUCH_NODE )
+                    {
+                        current.value[i] = propertyValue;
+                        current.nodeId[i] = nodeId;
+                        if ( i == entrySetSize - 1 )
+                        {
+                            current.next = new EntrySet( entrySetSize );
+                        }
+                        break scan;
+                    }
+                    else if ( property.valueEquals( value ) )
+                    {
+                        throw new IndexEntryConflictException( current.nodeId[i], nodeId, value );
+                    }
+                }
+                current = current.next;
+            }
+            while ( current != null );
+        }
+
+        private EntrySet bucketEntrySet( int hashCode, int entrySetSize )
+        {
+            int bucket = Math.abs( hashCode ) % numberOfBuckets;
+            EntrySet current = actualValues[bucket];
+            if ( current == null )
+            {
+                current = new EntrySet( entrySetSize );
+                actualValues[bucket] = current;
+            }
+            return current;
+        }
+
+        private static class EntrySet
```

Comment:
A small javadoc about this guy being a linked list where each entry contain `nodeId`:`propertyValue` pairs would be nice.